pkg/cluster/fixssh.go

```diff
@@ -124,52 +153,122 @@ func (m *manager) removeBackendPoolsFromNIC(ctx context.Context, resourceGroup,
 		ilbBackendPool = infraID + "-internal-controlplane-v4"
 	case api.ArchitectureVersionV2:
 		ilbBackendPool = infraID
 	}
 
-	sshID := fmt.Sprintf("%s/backendAddressPools/ssh-%d", *lb.ID, i)
-	ilbID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, ilbBackendPool)
+	sshBackendPoolID := fmt.Sprintf("%s/backendAddressPools/ssh-%d", *lb.ID, i)
+	ilbBackendPoolID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, ilbBackendPool)
 
 	updateSSHPool := true
 	updateILBPool := true
 
-	for _, p := range *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools {
-		if strings.EqualFold(*p.ID, sshID) {
-			updateSSHPool = false
-		}
-		if strings.EqualFold(*p.ID, ilbID) {
-			updateILBPool = false
+	ipc := (*nic.InterfacePropertiesFormat.IPConfigurations)[0]
+	if ipc.LoadBalancerBackendAddressPools == nil {
+		emptyBackendAddressPool := make([]mgmtnetwork.BackendAddressPool, 0)
+		(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = &emptyBackendAddressPool
+	} else {
+		for _, p := range *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools {
+			if strings.EqualFold(*p.ID, sshBackendPoolID) {
+				updateSSHPool = false
+			}
+			if strings.EqualFold(*p.ID, ilbBackendPoolID) {
+				updateILBPool = false
+			}
 		}
 	}
 
 	if updateSSHPool {
-		m.log.Printf("Adding NIC %s to Internal Load Balancer SSH Backend Address Pool %s", nicName, sshID)
+		m.log.Printf("Adding NIC %s to Internal Load Balancer SSH Backend Address Pool %s", nicName, sshBackendPoolID)
 		*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = append(*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools, mgmtnetwork.BackendAddressPool{
-			ID: &sshID,
+			ID: &sshBackendPoolID,
 		})
 	}
 
 	if updateILBPool {
-		m.log.Printf("Adding NIC %s to Internal Load Balancer API Address Pool %s", nicName, ilbID)
+		m.log.Printf("Adding NIC %s to Internal Load Balancer API Address Pool %s", nicName, ilbBackendPoolID)
 		*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = append(*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools, mgmtnetwork.BackendAddressPool{
-			ID: &ilbID,
+			ID: &ilbBackendPoolID,
 		})
 	}
 
 	return updateSSHPool || updateILBPool
 }
 
+func (m *manager) updateV1ELBAddressPool(ctx context.Context, nic *mgmtnetwork.Interface, nicName string, resourceGroup string, infraID string) error {
+	if nic.InterfacePropertiesFormat.IPConfigurations == nil || len(*nic.InterfacePropertiesFormat.IPConfigurations) == 0 {
+		return fmt.Errorf("unable to update NIC as there are no IP configurations for %s", nicName)
+	}
+
+	var updateV1ELBPool bool
+	lb, err := m.loadBalancers.Get(ctx, resourceGroup, infraID, "")
+	elbBackendPoolID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, infraID)
+	lbpool := *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools
+	var index int
+	for i, p := range lbpool {
+		if strings.EqualFold(*p.ID, elbBackendPoolID) {
+			updateV1ELBPool = true
+			index = i
+		}
+	}
+
+	if updateV1ELBPool {
```

Comment:
IMO we should not use 'tracking booleans' and do the work when we need to / the soonest we know which nic needs to be removed. I think we can also avoid having to track the index and avoid possible index out of range errors if we create a new slice and only insert the NICs we want to keep.

I think something like this would work, what do you think?

```golang
	if nic.InterfacePropertiesFormat.IPConfigurations == nil || len(*nic.InterfacePropertiesFormat.IPConfigurations) == 0 {
		return fmt.Errorf("unable to update NIC as there are no IP configurations for %s", nicName)
	}

	lb, err := m.loadBalancers.Get(ctx, resourceGroup, infraID, "")
	if err != nil {
		return err
	}
	elbBackendPoolID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, infraID)
	currentPool := *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools
	newPool := make([]mgmtnetwork.BackendAddressPool, len(currentPool))
	for _, pool := range currentPool {
		if strings.EqualFold(*pool.ID, elbBackendPoolID) {
			m.log.Printf("Removing NIC %s from Public Load Balancer API Address Pool %s", nicName, elbBackendPoolID)
		} else {
			newPool = append(newPool, pool)
		}
	}

	(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = &newPool
	m.log.Printf("Updating Network Interface %s", nicName)
	return m.interfaces.CreateOrUpdateAndWait(ctx, resourceGroup, nicName, *nic)
```

Reply:
Made changes as per the suggestions.